src/JoyOI.Monitor/Controllers/ChartController.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Http;
+using MySql.Data.MySqlClient;
+
+using JoyOI.Monitor.Models;
+using JoyOI.Monitor.Lib;
+using System.Drawing;
+
+namespace JoyOI.Monitor.Controllers
+{
+    public class ChartController : BaseController
+    {
+
+        [NonAction]
+        public async Task<Chart> GetChartData(
+            string datasource,
+            string sql,
+            ChartScaling scale,
+            Func<List<Dictionary<string, object>>, Chart> proc_rows
```

Comment:
Please don't use implementation type as argument type. You should use Func<IEnumerable<IDictionary<string, object>>, Chart> instead of the current one.